File: src/db-toolkit-ui/resources/backend/_internal/operations/session_manager.py
"""Session state management."""

import json
from pathlib import Path
from typing import Dict, Optional

from core.models import DatabaseConnection
from core.storage import ConnectionStorage
from utils.logger import logger


class SessionManager:
    """Manages session state and settings."""

    def __init__(self):
        """Initialize session manager."""
        self.session_file = Path.home() / ".db-toolkit" / "session.json"
        self.session_file.parent.mkdir(parents=True, exist_ok=True)
        self.storage = ConnectionStorage()

    async def save_session(
        self, active_connection_ids: list[str], last_active: Optional[str] = None
    ) -> bool:
        """Save current session state."""
        try:
            session_data = {
                "active_connection_ids": active_connection_ids,
                "last_active_connection": last_active,
                "settings": {},
            }

            with open(self.session_file, "w") as f:
                json.dump(session_data, f, indent=2)

            return True
        except Exception as e:
            logger.error(f"Failed to save session: {str(e)}")
            return False

    async def load_session(self) -> Dict:
        """Load saved session state."""
        try:
            if self.session_file.exists():
                with open(self.session_file, "r") as f:
                    return json.load(f)
        except Exception as e:
            logger.error(f"Failed to load session: {str(e)}")

        return {"active_connection_ids": [], "last_active_connection": None, "settings": {}}

    async def clear_session(self) -> bool:
        """Clear session state."""
        try:
            if self.session_file.exists():
                self.session_file.unlink()
            return True
        except Exception as e:
            logger.error(f"Failed to clear session: {str(e)}")
            return False

    async def get_restorable_connections(self) -> list[DatabaseConnection]:
        """Get connections that can be restored from session."""
        session = await self.load_session()
        connections = []

        for conn_id in session.get("active_connection_ids", []):
            conn = await self.storage.get_connection(conn_id)
            if conn:
                connections.append(conn)

        return connections


session_manager = SessionManager()

File: src/db-toolkit/electron/backend/utils/session-manager.js
/**
 * Session state management.
 */

const fs = require('fs').promises;
const path = require('path');
const os = require('os');
const { connectionStorage } = require('./connection-storage');
const { logger } = require('./logger');

class SessionManager {
  constructor() {
    this.sessionFile = path.join(os.homedir(), '.db-toolkit', 'session.json');
  }

  async ensureDir() {
    const dir = path.dirname(this.sessionFile);
    await fs.mkdir(dir, { recursive: true });
  }

  async saveSession(activeConnectionIds, lastActive = null) {
    try {
      await this.ensureDir();
      const sessionData = {
        active_connection_ids: activeConnectionIds,
        last_active_connection: lastActive,
        settings: {},
      };

      await fs.writeFile(this.sessionFile, JSON.stringify(sessionData, null, 2), 'utf-8');
      return true;
    } catch (error) {
      logger.error('Failed to save session:', error);
      return false;
    }
  }

  async loadSession() {
    try {
      const data = await fs.readFile(this.sessionFile, 'utf-8');
      return JSON.parse(data);
    } catch (error) {
      if (error.code !== 'ENOENT') {
        logger.error('Failed to load session:', error);
      }
      return { active_connection_ids: [], last_active_connection: null, settings: {} };
    }
  }

  async clearSession() {
    try {
      await fs.unlink(this.sessionFile);
      return true;
    } catch (error) {
      if (error.code !== 'ENOENT') {
        logger.error('Failed to clear session:', error);
        return false;
      }
      return true;
    }
  }

  async getRestorableConnections() {
    const session = await this.loadSession();
    const connections = [];

    for (const connId of session.active_connection_ids || []) {
      const conn = await connectionStorage.getConnection(connId);
      if (conn) {
        connections.push(conn);
      }
    }

    return connections;
  }
}

const sessionManager = new SessionManager();

module.exports = { SessionManager, sessionManager };
